README.md
# TinNS (TinNS is not a Neocron Server)

This is a mirror of the official svn repo hosted at http://www.linux-addicted.net/
I worked on this a couple of years ago and I thought it might be nice to have a mirror here if anyone is interested.

# What is it?

TinNS is a reverse engineered server for Neocron 1, a by now outdated cyberpunk-genre MMORPG, see: http://en.wikipedia.org/wiki/Neocron

server/CHANGELOG
<GPL>
	TinNS (TinNS is not a Neocron Server)
	Copyright (C) 2005 Linux Addicted Community
	maintainer Akiko <akiko@gmx.org>

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License
	as published by the Free Software Foundation; either version 2
	of the License, or (at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
	02110-1301, USA.
</GPL>

Developers:
- Akiko
- bakkdoor
- Namikon
- Hammag

> Recent changes
====================
Namikon; 31.05.2006 22:12 GMT+1
----------
<General changes>
- Introduced CHANGELOG file
- Added version.h to enable a way to create release tarballs/binaries
- Fixed a dozen of typos and bugs from the additions of the last 2 days
- "Clean build"; Should compile without any error now. (Note: Not "work", "compile" :) )


<Source changes>
[GAME]
globals.cpp :: ShutdownTinNS()
  - Renamed to Shutdown()
  - Added svn version output to startup procedure
  - Fixed minor typos/console outputs
config.cpp :: LoadOptions()
  - Changed configfile, now loads data from /conf/gameserver.cfg
  - Added full 2 DB support (was: "sql_*" now: "info_sql_*" and "game_sql_*")

[INFO]

[PATCH]
globals.cpp :: ShutdownTinNS()
  - Renamed to Shutdown()

[ALL]
misc.cpp :: GetSVNRev()
  - Added handling for version.h defines
====================
Hammag; 06.07.2006 02:17 GMT+1
----------
<General changes>
- Put UDP sockets in non-blocking mode.
- Checked and modified all socket related stuff to work in non-blocking mode.
- Lowered CPU use from 100% to ~0% by changing from sched_yield() to I/O wait sleep using Select.
  This avoid runingthe main loop when there's nothing to do (some more improvements are to come
  on this topic).
- The socket Select now waits for any tcp OR UDP socket incoming data.
- Fixed some pointer issues in netcode.


<Source changes>
[GAME]
 
[INFO]
misc.h
 - added declaration of utility function IPlongToString()
misc.cpp :: IPlongToString()
 - implemented function to convert IP address in u32 format to string
infoserver.cpp :: HandleServerList()
 - now use IPlongToString() to display server IP on Console

[PATCH]

[ALL]
patchserver.cpp, infoserver.cpp, gameserver.cpp :: Start()
 - set 10 msec timeout for socket select, which sets the max wait time
  for the main loop when no activity.

main.cpp
 - commented out sched_yield() in main loop, as it is not needed anymore.

main.h
 - moved-up tinns-specific types include to ready them for netcode compilation

[NETCODE]
connection-tcp.cpp :: ConnectionTCP()
 - moved-in non-blocking setting from ServerSocket::getTCPConnection()
    for better class coherency
connection-tcp.cpp :: update()
 - modified to deal correctly with would-block socket writes.
 - fixed a pointer and size issue in socket recv

connection-udp.h :: class
 - added private member m_ServerSocket
 - added ServerSocket parameter in constructor prototype
 
connection-udp.cpp :: ConnectionUDP()
 - modified to set m_ServerSocket from constructor parameters
connection-udp.cpp :: ~ConnectionUDP()
 - added socket removal from fd sets by calling m_ServerSocket->delSocketFromSet
connection-udp.cpp :: update()
 - modified to check m_ServerSocket->isDataAvailable() before trying to read from socket
 - fixed a pointer and size issue in socket recv

serversocket.h :: class
 - added settimeout public method
 - added fd_set m_MainSetUDP private member
 - added fd_set m_MainSetGlobal private member (= m_MainSetTCP + m_MainSetUDP)
 
serversocket.cpp :: ServerSocket()
 - moved m_TimeOut initialization in constructor
 - added m_MainSetUDP and m_MainSetGlobal init
serversocket.cpp :: settimeout()
 - implemented method to set m_TimeOut to permit tuning
serversocket.cpp :: open()
 - added m_MainSetUDP and m_MainSetGlobal init
 - added m_ListenerTCP to m_MainSetGlobal too
serversocket.cpp :: update()
 - now m_MainSetGlobal is copied to m_ReadSetTCP set,
 - added a local fdMax for select
 - use a temp timeout structure in select as select doc says select implementation
    for Linux could change this value
serversocket.cpp :: getUDPConnection()
 - now pass this pointer to new ConnectionUDP constructor
 - new UDP sockets are added to m_MainSetUDP and m_MainSetGlobal,
    and kept track of in m_FdMaxUDP
serversocket.cpp :: getTCPConnection()
 - fixed a struct length issue (?) in accept
 - now new TCP sockets are added to m_MainSetGlobal too
 - moved non-blocking setting to ConnectionTCP::ConnectionTCP() for better class coherency
serversocket.cpp :: delSocketFromSet()
 - added removal of sockfd from m_MainSetUDP and m_MainSetGlobal too
 
====================
Hammag; 30.07.2006 02:17 GMT+1
----------
<General changes>
 - Baseline fixed, full zoning now really works with multiplayer
 - Multiuser now works. Multiple login + mutual viewing.
    Now a char displays as expected ingame for himself as for others.(some movement/positionning glitches remain)
 - Warning: code for using items hasn't been fixed, so every thing 2nd+ user does will be
     attriuted to 1st user and provoque OOO flood.
     So don't try using items, except for the first user.
 - added @skin command. Reminder use by @skin, little mode description in commands.cpp header

<Source changes>
[GAME]
 - Many changes in many files
 - amongst other, rewrited baseline.inc.cpp with new PMessage object as an exemple
 - added PItem & PInventory classes

[NETCODE]

server/src/game/main.h
/*
	TinNS (TinNS is not a Neocron Server)
	Copyright (C) 2005 Linux Addicted Community
	maintainer Akiko <akiko@gmx.org>

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License
	as published by the Free Software Foundation; either version 2
	of the License, or (at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
	02110-1301, USA.
*/



/*
	main.h - main include file, contains all needed includes and important definitions

	MODIFIED: 31 Aug 2005 Akiko
	REASON: - updating def file definitions to match the changes in gamedefs.cpp
		- added define for charfiles path
		- added define for database path
	MODIFIED: 26 Sep 2005 Akiko
	REASON: - reformated code
		- added GPL
		- added infoserver default port as define
		- added my threading model (mutex.h, semaphore.h, thread.h)
		- added basic includes pthread.h, semaphore.h and errno.h
	MODIFIED: 28 Sep 2005 Akiko
	REASON:	- added define for game server port
	MODIFIED: 30 Nov 2005 Akiko
	REASON: - added chat.h for the chat part from Namikon
	MODIFIED: 02 Dec 2005 Akiko
	REASON: - commented out mutex.h, thread.h, semaphore.h
	MODIFIED: 22 Dec 2005 Namikon/bakkdoor
	REASON: - Added commands.h, skill.h, clientmanager.h
    MODIFIED: 23 Dec 2005 bakkdoor
	REASON: - Added <csignal> for main.cpp -> catch strg-c and shutdown nicely
    MODIFIED: 25 Dec 2005 Namikon
	REASON: - Added mysql.h, sql.h for MySQL support
    MODIFIED: 01 Jan 2006 Namikon
	REASON: - Moved skill.h before chars.h (char.h needs skill.h now)
    MODIFIED: 30 May 2006 Namikon
	REASON: - Removed all useless includes to complete the server splitup; Also renamed tinns.h to main.h
    MODIFIED: 6 Jul 2006 Hammag
	  REASON: - moved include "types.h" before include "../netcode/main.h" to permit compile
    MODIFIED: 10 Jul 2006 Hammag
	  REASON: - added inventory.h
	  REASON: - added item.h
    MODIFIED: 26 Jul 2006 Hammag
	  REASON: - removed	#define GAME_PORT which is not used anymore (now in config file)

*/

#ifndef MAIN_H
#define MAIN_H

//#include "version.h"

//basic includes
#include "external.h"

//tinns includes
#include "types.h"
#include "netcode.h"
/*
#include "../gamemonkey/gmMachine.h"
#include "../gamemonkey/gmCall.h"
*/
#include "console.h"

// MySQL Support
#include "mysql.h"
#include "sql.h"

#include "config.h"
#include "filesystem.h"
#include "defparser.h"

#include "skill.h"
#include "chars.h"
#include "accounts.h"
#include "client.h"
#include "server.h"
#include "misc.h"
#include "gameserver.h"
#include "globals.h"
#include "defs.h"
#include "zoning.h"
#include "item.h"
#include "inventory.h"
#include "worldactors.h"
#include "npc.h"
#include "terminal.h"

#include "chat.h"
#include "commands.h"
#include "clientmanager.h"

using namespace std;

#endif

